java-backend/src/main/java/org/kframework/backend/java/symbolic/InitializeRewriter.java

```diff
@@ -331,13 +331,17 @@ public ProcessProofRules invoke(GlobalContext rewritingContext, long initCounter
             }
 
             public org.kframework.backend.java.kil.Rule evaluateRule(org.kframework.backend.java.kil.Rule rule) {
+                termContext.setTopConstraint(null);
                 //We need this ConsTerm only to evaluate the constraint. That's why we use an empty first argument.
                 ConstrainedTerm constraintHolder = new ConstrainedTerm(
                         ConjunctiveFormula.of(termContext.global()),
                         rule.getRequires(),
                         termContext).expandPatterns(true);
 
                 ConjunctiveFormula constraint = constraintHolder.constraint();
+                termContext.setTopConstraint(constraint);
+                //simplify the constraint in its own context, to force full evaluation of terms.
+                constraint = constraint.simplify(termContext);
```

Comment:
what will happen if you simply the constraint before `setTopConstraint`?

Reply:
Then it won't get fully simplified. In the current implementation an equality in constraint is simplified in the context of all other equalities. If constraint is null, then it will be simplified with no constraint.

Reply:
Essentially in this class constraint is simplified twice - once as part of expandPatterns(), in the presence of null context, the 2nd one - in the presence of itself. 2 steps are required because in the 1st step substitutions are calculated, and then you can use the constraint with proper substitutions to simplify itself once again.

All other changes are just refactoring/removing redundant code.

Reply:
i mean, what's the problem with the 1st simplification under the non-null `constraint`?

Reply:
Initial constraint doesn't contain substitutions. First simplification generates those substitutions. Then we need another simplification to simplify in the presence of substitutions. Code is really cryptic and if you really want to understand all implications you have to look into many other classes not touched by this change.

But one thing you can probably agree - 2 simplifications instead of one cannot break things:)